lighthouse-core/audits/third-party-facades.js

```diff
@@ -0,0 +1,241 @@
+  /**
+   * @description Template for an entry in the "Product" column for a third party product which is in the video category.
+   * @example {YouTube Embedded Player} productName
+   */
+  categoryVideo: '{productName} (Video)',
+  /**
+   * @description Template for an entry in the "Product" column for a third party product which is in the customer success category.
+   * @example {Intercom Widget} productName
+   */
+  categoryCustomerSuccess: '{productName} (Customer Success)',
+  /**
+   * @description Template for an entry in the "Product" column for a third party product which is in the marketing category.
+   * @example {Drift Live Chat} productName
+   */
+  categoryMarketing: '{productName} (Marketing)',
+  /**
+   * @description Template for an entry in the "Product" column for a third party product which is in the social category.
+   * @example {Facebook Messenger Customer Chat} productName
+   */
+  categorySocial: '{productName} (Social)',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+const CATEGORY_UI_MAP = new Map([
+  ['video', UIStrings.categoryVideo],
+  ['customer-success', UIStrings.categoryCustomerSuccess],
+  ['marketing', UIStrings.categoryMarketing],
+  ['social', UIStrings.categorySocial],
+]);
+
+/** @typedef {import("third-party-web").IEntity} ThirdPartyEntity */
+/** @typedef {import("third-party-web").IProduct} ThirdPartyProduct*/
+/** @typedef {import("third-party-web").IFacade} ThirdPartyFacade*/
+
+/** @typedef {{
+ *  product: ThirdPartyProduct,
+ *  cutoffTime: number,
+ *  urlSummaries: Map<string, ThirdPartySummary.Summary>
+ * }} FacadableProductSummary
+ */
+
+class ThirdPartyFacades extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'third-party-facades',
+      title: str_(UIStrings.title),
+      failureTitle: str_(UIStrings.failureTitle),
+      description: str_(UIStrings.description),
+      requiredArtifacts: ['traces', 'devtoolsLogs', 'URL'],
+    };
+  }
+
+  /**
+   * @param {Map<string, ThirdPartySummary.Summary>} byURL
+   * @param {ThirdPartyEntity | undefined} mainEntity
+   * @return {FacadableProductSummary[]}
+   */
+  static getFacadableProductSummaries(byURL, mainEntity) {
+    /** @type {Map<string, Map<string, FacadableProductSummary>>} */
+    const entitySummaries = new Map();
+
+    for (const [url, urlSummary] of byURL) {
+      const entity = thirdPartyWeb.getEntity(url);
+      if (!entity || thirdPartyWeb.isFirstParty(url, mainEntity)) continue;
+
+      const productSummaries = entitySummaries.get(entity.name) || new Map();
+      const product = thirdPartyWeb.getProduct(url);
+      if (product && product.facades && product.facades.length) {
+        // Record new url if product has a facade.
+
+        /** @type {FacadableProductSummary} */
+        const productSummary = productSummaries.get(product.name) || {
+          product,
+          cutoffTime: Infinity,
```

Comment:
should this be named something with `start` in it? Since we are only going to associate things with it _after_ this time, but `cutoff` sounds like the end of something (to me, at least :)

Reply:
I thought that might get mixed up with `startTime` since `cutoff` is the `endTime`-ish of some request, maybe `beginning` works better?